File: python-sdk/src/aryacrypt/kdf.py
import hashlib
import re

from .constants import KEY_LENGTH_BYTES, PBKDF2_HASH, PBKDF2_ITERATIONS, SALT_LENGTH_BYTES
from .errors import AryaCryptError


def derive_key(password_material: bytes, salt: bytes, iterations: int = PBKDF2_ITERATIONS) -> bytes:
    if not isinstance(password_material, bytes):
        raise AryaCryptError("KDF password material must be bytes.")
    if not isinstance(salt, bytes):
        raise AryaCryptError("Salt must be bytes.")
    if len(salt) != SALT_LENGTH_BYTES:
        raise AryaCryptError(f"Salt must be exactly {SALT_LENGTH_BYTES} bytes.")
    if iterations < 100_000:
        raise AryaCryptError("PBKDF2 iterations must be at least 100000.")

    return hashlib.pbkdf2_hmac(
        hash_name=PBKDF2_HASH,
        password=password_material,
        salt=salt,
        iterations=iterations,
        dklen=KEY_LENGTH_BYTES,
    )

File: node-sdk/src/kdf.ts
import { pbkdf2 as pbkdf2Cb } from "node:crypto";
import { promisify } from "node:util";
import {
  KEY_LENGTH_BYTES,
  PBKDF2_ITERATIONS,
  SALT_LENGTH_BYTES,
} from "./constants.js";
import { AryaCryptError } from "./errors.js";

const pbkdf2 = promisify(pbkdf2Cb);

export async function deriveKey(
  passwordMaterial: Buffer,
  salt: Buffer,
  iterations: number = PBKDF2_ITERATIONS
): Promise<Buffer> {
  if (!Buffer.isBuffer(passwordMaterial)) {
    throw new AryaCryptError("KDF password material must be a Buffer.");
  }
  if (!Buffer.isBuffer(salt) || salt.length !== SALT_LENGTH_BYTES) {
    throw new AryaCryptError(`Salt must be exactly ${SALT_LENGTH_BYTES} bytes.`);
  }
  if (iterations < 100_000) {
    throw new AryaCryptError("PBKDF2 iterations must be at least 100000.");
  }
  return pbkdf2(passwordMaterial, salt, iterations, KEY_LENGTH_BYTES, "sha256");
}
